build.assets/tooling/cmd/notarize-apple-binaries/args.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"debug/macho"
+	"fmt"
+
+	"github.com/gravitational/kingpin"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+const binaryArgName string = "binaries"
+
+type Args struct {
+	logLevel      *string
+	logJSON       *bool
+	AppleUsername *string
+	ApplePassword *string
+	BinaryPaths   *[]string
+}
+
+func NewArgs() *Args {
```

Comment:
I'd still tweak this a bit. IMO `kingping.Flag` calls should literally be in `main()`. The way this is written you can really only call `NewArgs` once and it has to be before flags are parsed, which is awkward.

I would make these changes to eliminate the tight coupling between what is an arbitrary struct and flag parsing.

1. Remove `NewArgs`. You don't need it.
2. Remove the pointer types from `Args`'s fields. You don't need them, and we shouldn't let the API of a flag parsing library dictate the structure of other types.

Then in main, something like this should suffice.

```
func main() {
    var args Args
    kingpin.Flag("apple-username", "...").Envar("...").StringVar(&args.AppleUsername)
    kingpin.Flag("apple-password", "..").Envar("..").StringVar(&args.ApplePassword)
    kingping.Arg("binary-paths", "..").Required().Action(..).ExistingFilesVar(&args.BinaryPaths)
    kingpin.Parse()
```

With this approach, your `Args` struct is just plain old data. In this case, it's set from CLI flags, but it doesn't have to be. (You could hard code values for testing, etc)

In fact, now that it's not coupled to the CLI a better name for `Args` might be `Config` :-)

Reply:
I have updated this as requested. That being said I am not a huge fan of this approach as (in my opinion) it puts too much logic in a single function that should only handle very high level logic (`main`). 

Additionally, because the validation functions (`binaryArgValidatiorAction` and `verifyFileIsAppleBinary`) are only used when setting up the flags/args (in `main.go`), I have moved them to this file as well. That only leaves the struct in `args.go`, which is now only used in `main.go`, so I moved it to as well. As a result there is now a lot of additional "business logic" code in `main.go`, rather than having `main.go` simply connect high level business logic.

Lastly this differs from the other tools in `build.assets/tooling` where all the argument parsing and validation is handled in an `args.go` or similar file(s) outside of `main.go`.

I will defer to your expertise but personally I think it is much cleaner to have the logic live entirely in a separate file.